flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/statistics/basicstatistic/MultivariateGaussian.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.statistics.basicstatistic;
+
+import org.apache.flink.ml.common.linalg.BLAS;
+import org.apache.flink.ml.common.linalg.DenseMatrix;
+import org.apache.flink.ml.common.linalg.DenseVector;
+import org.apache.flink.ml.common.linalg.SparseVector;
+import org.apache.flink.ml.common.linalg.Vector;
+
+import com.github.fommil.netlib.LAPACK;
+import org.netlib.util.intW;
+
+/**
+ * This class provides basic functionality for a Multivariate Gaussian (Normal) Distribution.
+ */
+public class MultivariateGaussian {
+
+	private static final LAPACK LAPACK_INST = LAPACK.getInstance();
+	private static final double EPSILON;
+
+	static {
+		double eps = 1.0;
+		while ((1.0 + (eps / 2.0)) != 1.0) {
+			eps /= 2.0;
+		}
+		EPSILON = eps;
+	}
+
+	private final DenseVector mean;
+	private final DenseMatrix cov;
+
+	private DenseMatrix rootSigmaInv;
+	private double u;
+
+	// data buffers for computing pdf
+	private DenseVector delta;
+	private DenseVector v;
+
+	/**
+	 * The constructor.
+	 *
+	 * @param mean The mean vector of the distribution.
+	 * @param cov  The covariance matrix of the distribution.
+	 */
+	public MultivariateGaussian(DenseVector mean, DenseMatrix cov) {
+		this.mean = mean;
+		this.cov = cov;
+		this.delta = DenseVector.zeros(mean.size());
+		this.v = DenseVector.zeros(mean.size());
+		calculateCovarianceConstants();
+	}
+
+	/**
+	 * Returns density of this multivariate Gaussian at given point, x.
+	 */
+	public double pdf(Vector x) {
+		return Math.exp(logpdf(x));
+	}
+
+	/**
+	 * Returns the log-density of this multivariate Gaussian at given point, x.
+	 */
+	public double logpdf(Vector x) {
+		int n = mean.size();
+		System.arraycopy(mean.getData(), 0, delta.getData(), 0, n);
+		BLAS.scal(-1.0, delta);
+		if (x instanceof DenseVector) {
+			BLAS.axpy(1., (DenseVector) x, delta);
+		} else if (x instanceof SparseVector) {
+			BLAS.axpy(1., (SparseVector) x, delta);
+		}
+		BLAS.gemv(1.0, rootSigmaInv, false, delta, 0., v);
+		return u - 0.5 * BLAS.dot(v, v);
+	}
+
+	/**
+	 * Compute distribution dependent constants.
+	 *
+	 * <p>rootSigmaInv = D^(-1/2)^ * U.t, where sigma = U * D * U.t .
```

Comment:
much more informatively: 
```
 * <p><ul>
 * <li> sigma = U * D * U.t
 * <li> inv(Sigma) = U * inv(D) * U.t = = (D^{-1/2}^ * U.t).t * (D^{-1/2}^ * U.t)
 * <li> sqrt(inv(Sigma)) = D^(-1/2)^ * U.t
 * </ul></p>
```

since the word `rootSigmaInv` is not intuitively understandable to me. (especially "root" doesn't mean square-root)

Reply:
Thanks, we refined the JavaDoc according to your advice.